### DISCUSSION TITLE
Combining residual block and layer norm fusion

### DISCUSSION THREAD

**Discussion by masahi** [NVIDIA/cutlass#874] [Q&A] [answered]
Hi, I'm looking at optimization for stable diffusion. One very common pattern is gemm -> residual add -> layer norm. In CUTLASS, both residual block and layer norm fusion are supported in isolation (via https://github.com/NVIDIA/cutlass/blob/277bd6e5379e0c1e1eb64db1a654b30e1efddc8e/include/cutlass/gemm/kernel/gemm_with_fused_epilogue.h#L70 and https://github.com/NVIDIA/cutlass/blob/66d9cddc832c1cdc2b30a8755274f7f74640cfe6/include/cutlass/gemm/kernel/gemm_layernorm_mainloop_fusion.h#L63), but since the two kernels don't seem to "compose" each other, I don't see a way to realize residual + layer norm fusion simultaneously.

Am I missing something? This should be a very common problem people encounter in practice... so I hope there is a good solution. Or is this also what the V3 API can solve?

@hwu36 @thakkarV

**Comment by masahi** [comment, marked as answer]
Looking more closely at how layer norm fusion is supported, I'm assuming that layer norm fusion in cutlass is intended to be used for gemm -> layer norm -> gemm fusion. Since the residual add happens before layer norm, I think what I need to do is to replace the first gemm implementation https://github.com/NVIDIA/cutlass/blob/master/examples/37_gemm_layernorm_gemm_fusion/gemm_with_epilogue_visitor.h with the one that supports residual addition. 

The situation seems much better now.

**Reply by masahi** [reply]
Actually it looks even simpler: I think I just need to modify https://github.com/NVIDIA/cutlass/blob/master/examples/37_gemm_layernorm_gemm_fusion/gemm_with_layernorm.h#L525 to support residual addition? This is the code for first GEMM epilogue.

**Reply by jackkosaian** [reply]
Hi, @masahi. Yes, I believe you can change `ElementFunctorOp` [here](https://github.com/NVIDIA/cutlass/blob/af332d4aa9ee19f901f0ceac442078fb04b350fb/examples/37_gemm_layernorm_gemm_fusion/gemm_layernorm.cu#L283-L288) to be `LinearCombinationResidualBlock` (defined [here](https://github.com/NVIDIA/cutlass/blob/af332d4aa9ee19f901f0ceac442078fb04b350fb/include/cutlass/epilogue/thread/linear_combination_residual_block.h#L65)).

You would also need to change the `Arguments` structures (e.g., [here](https://github.com/NVIDIA/cutlass/blob/af332d4aa9ee19f901f0ceac442078fb04b350fb/examples/37_gemm_layernorm_gemm_fusion/gemm_with_layernorm.h#L305-L342)) so that you could pass in the residual tensor.

**Reply by MingZwhy** [reply]
Hi, @jackkosaian , If I change ElementFunctorOp into LinearCombinationResidualBlock, will the EpilogueFunctorOp of the second gemm also be changed? If so, can I set the EpilogueFunctorOp for two gemm splitly? For example, set the EpilogueFunctorOp of the first gemm to be residual and the EpilogueFunctorOp of the second gemm to be relu. Thanks for your advise!

**Reply by jackkosaian** [reply]
@MingZwhy , you should be able to do this. You'll just need to do some plumbing so that the `EpilogueFunctorOp` passed to `GemmMainloopFusion` ([here](https://github.com/NVIDIA/cutlass/blob/9385141f19d87d9fb4c24fb29b65d0e8facd55a7/examples/37_gemm_layernorm_gemm_fusion/gemm_with_layernorm.h#L828-L842)) uses ReLU.
